migrate.py

```diff
@@ -0,0 +1,217 @@
+#!/usr/bin/env python3
+from typing import List, Optional, Tuple, Dict, Any
+
+import os
+import argparse
+import subprocess
+import datetime
+import json
+
+ZFS_SNAPSHOTDIR = '.zfs/snapshot'
+
+SNAPSHOT_TAG = "snapshot="
+
+DEBUG = False
+
+
+def _run(command: str, input: Optional[str] = None, void_stderr: bool = False) -> None:
+    other_args = dict()
+    if void_stderr and not DEBUG:
+        other_args["stderr"] = subprocess.DEVNULL
+    subprocess.run(command, shell=True, text=True, input=input, **other_args)
+
+
+def _eval(command: str, input: Optional[str] = None, void_stderr: bool = False) -> str:
+    other_args = dict()
+    if void_stderr and not DEBUG:
+        other_args["stderr"] = subprocess.DEVNULL
+    return subprocess.run(command, shell=True, text=True, stdout=subprocess.PIPE, input=input, **other_args).stdout
+
+
+class Backuper:
+
+    def __init__(self,
+                 restic_repo_prefix: str,
+                 zfs_dataset_common_prefix: str,
+                 restic_password_file: str):
+        self.restic_repo_prefix: str = restic_repo_prefix.rstrip("/")
+        self.zfs_dataset_common_prefix: str = zfs_dataset_common_prefix
+        self.restic_password_file: str = restic_password_file
+
+    def _restic_cmd(self, restic_repo: str, restic_command: str, flags: List[str] = []) -> str:
+        initial_args = ["-r", restic_repo, "--password-file", self.restic_password_file, restic_command]
+        args = initial_args + flags
+        arg_string = " ".join([f"'{arg}'" for arg in args])
+        return f"restic {arg_string}"
+
+    def _get_dataset_snapshots(self, dataset_name: str) -> List[str]:
+        lines = _eval(f"sudo zfs list -Hp -o name -t snapshot '{dataset_name}'")
+        snapshots = [line.split("@")[-1] for line in lines.split("\n") if len(line) > 0]
+        return snapshots
+
+    def _get_snapshot_tag(self, datum: Dict[str, Any]) -> str:
+        tags = datum["tags"]
+        tag: str
+        for tag in tags:
+            if tag.startswith(SNAPSHOT_TAG):
+                return tag[len(SNAPSHOT_TAG):]
+        raise Exception("Snapshot does not have a valid snapshot tag.")
+
+    def _get_snapshots_in_restic(self, restic_repo: str) -> Dict[str, str]:
+        json_data = _eval(self._restic_cmd(restic_repo, "snapshots", ["--json"]))
+        data = json.loads(json_data)
+        return {self._get_snapshot_tag(datum): datum["short_id"] for datum in data}
```

Comment:
I guess short IDs can occasionally conflict. Is there a full ID which we can use instead?